packages/hir/src/lowering/types.candy

```diff
@@ -1 +1,103 @@
-// These impls make type system information available.
+use ast
+use compiler_utils
+use incremental
+
+use ...declarations
+use ...types
+use ..declarations
+
+/// These impls make type system information available.
+
+impl HirReflectionType: HasDeclarations {
+  fun typeUnsafeAstDeclarations(context: QueryContext<List<CompilerError>>): List<Any> {
+    List.empty<Any>()
+  }
+  public fun declarations(context: QueryContext<List<CompilerError>>): List<HirDeclaration> {
+    List.empty<HirDeclaration>()
+  }
+}
+
+// TODO(marcelgarus): Convert this into extension methods on Ast as soon as impls for types from
+// other packages work.
+fun astTypeToHirType(
+  context: QueryContext<List<CompilerError>>,
+  astType: AstType,
+  scope: HirDeclaration,
+): HirType {
+  query<HirType, List<CompilerError>>(
+    context,
+    "AstType.hir",
+    DataTuple2<AstType, HirDeclaration>(astType, scope) as Equals & Hash,
+    {
+      if (astType is AstNamedType) {
+        let nameParts = ((astType as AstNamedType).nameParts as Iterable<AstIdentifier>)
+          .map<String>({ it.value })
+
+        // Try to find the first part of the `nameParts`. Then, find children inside its
+        // declarations.
+        mut let candidates = nameParts.first().unwrap().resolveIdentifier(context, scope)
+        for namePart in nameParts.skip(1) {
+          candidates = ((candidates as Iterable<HirDeclaration>)
+            .expand<HirDeclaration>({
+              namePart.resolveIdentifier(context, it) as Iterable<HirDeclaration>
```

Comment:
When resolving `Foo Bar`, wouldn't this accept a `Bar` next to `Foo` if `Foo` itself doesn't contain a `Bar`? (As the resolver traverses parent scopes.)